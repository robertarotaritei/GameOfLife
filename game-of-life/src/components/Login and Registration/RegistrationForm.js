import React from 'react';
import InputField from './InputField';
import SubmitButton from './SubmitButton';
import { Redirect } from 'react-router-dom';
import UserStore from '../../stores/UserStore';
import { runInAction } from 'mobx';
import Typography from '@material-ui/core/Typography';

class RegistrationForm extends React.Component {

  constructor() {
    super();
    this.state = {
      username: '',
      password: '',
      passwordConfirmation: '',
    }

    runInAction(() => {
      UserStore.key = sessionStorage.getItem('key');
      UserStore.loading = false;
    });
  }

  setInputValue(property, val) {
    val = val.trim();
    if (val.length > 15) {
      return;
    }
    this.setState({
      [property]: val
    })
  }

  setRedirect = () => {
    this.setState({
      redirect: true
    })
  }
  renderRedirect = () => {
    if (this.state.redirect) {
      return <Redirect to='/login' />
    }
  }


  async doRegister() {
    if (!this.state.username) {
      return;
    }

    if (!this.state.password) {
      return;
    }

    if (!this.state.passwordConfirmation) {
      return;
    }

    runInAction(() => {
      UserStore.loading = true;
    });

    let error = '';
    error = this.state.password !== this.state.passwordConfirmation ? 'Passwords do not match' : '';
    let res = await fetch(`/credentials/user/verifyusername?username=${this.state.username}`)
    let result = await res.json();

    runInAction(() => {
      UserStore.loading = false;
    });

    error = result.title !== "Not Found" ? 'Username already exists' : error;
    if (error === '') {
      runInAction(() => {
        UserStore.loading = true;
      });

      var newUser = {
        username: this.state.username,
        password: this.state.password
      };
      fetch('/credentials/user', {
        method: 'POST',
        mode: 'cors',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(newUser)
      })

      alert("New user created");
      this.setRedirect();

      runInAction(() => {
        UserStore.loading = false;
      });
    }
    else {
      alert(error);
    }
  }

  render() {
    return (
      <div className="loginForm">
        {this.renderRedirect()}
        <div style={{ textAlign: 'center', color: '#17c5fa', marginTop: '25px' }}>
          <Typography variant='h5' style={{ fontWeight: 'bold' }}>
            Register
          </Typography>
        </div>
        <InputField
          className='username'
          type='text'
          placeholder='Username'
          value={this.state.username ? this.state.username : ''}
          onChange={(val) => this.setInputValue('username', val)}
        />
        <InputField
          className='password'
          type='password'
          placeholder='Password'
          value={this.state.password ? this.state.password : ''}
          onChange={(val) => this.setInputValue('password', val)}
        />
        <InputField
          className='repeatPassword'
          type='password'
          placeholder='Repeat Password'
          value={this.state.passwordConfirmation ? this.state.passwordConfirmation : ''}
          onChange={(val) => this.setInputValue('passwordConfirmation', val)}
        />
        <div style={{ textAlign: 'center' }}>
          <SubmitButton
            text='REGISTER'
            onClick={() => this.doRegister()}
          />
        </div>
        <div style={{ display: 'flex', marginTop: '30px', marginBottom: '20px' }}>
          <Typography variant='h5' style={{ fontWeight: 'bold' }}>
            Already have an account?
          </Typography>
          <div style={{ color: '#17c5fa' }}>
              <button className='button' onClick={this.setRedirect} >
                Log in
              </button>
          </div>
        </div>
      </div>
    );
  }
}

export default RegistrationForm;