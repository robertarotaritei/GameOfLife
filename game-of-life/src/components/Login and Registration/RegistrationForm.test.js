import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import RegistrationForm from './RegistrationForm';
import { shallow } from "enzyme";

test("RegistrationForm renders without crashing", () => {
  shallow(<RegistrationForm />);
});

test('Render RegistrationForm Text', () => {
  const { getByText } = render(<RegistrationForm />);
  const linkElement = getByText(/Already have an account/i);
  expect(linkElement).toBeInTheDocument();
});

test('Render RegistrationForm Log in Button', () => {
  const { getByText } = render(<RegistrationForm />);
  const linkElement = getByText(/Log in/i);
  expect(linkElement).toBeInTheDocument();
});

it('RegistrationForm renders correctly', () => {
  const tree = renderer
    .create(<RegistrationForm />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});