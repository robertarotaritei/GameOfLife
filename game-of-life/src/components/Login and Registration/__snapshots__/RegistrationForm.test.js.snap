// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm renders correctly 1`] = `
<div
  className="loginForm"
>
  <div
    style={
      Object {
        "color": "#355e3b",
        "marginTop": "30px",
        "textAlign": "center",
      }
    }
  >
    <h5
      className="MuiTypography-root MuiTypography-h5"
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Register
    </h5>
  </div>
  <div
    className="inputField"
    style={
      Object {
        "marginTop": "30px",
      }
    }
  >
    <input
      className="input"
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value=""
    />
  </div>
  <div
    className="inputField"
    style={
      Object {
        "marginTop": "30px",
      }
    }
  >
    <input
      className="input"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
  </div>
  <div
    className="inputField"
    style={
      Object {
        "marginTop": "30px",
      }
    }
  >
    <input
      className="input"
      onChange={[Function]}
      placeholder="Repeat Password"
      type="password"
      value=""
    />
  </div>
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <div
      className="submitButton"
      style={
        Object {
          "marginTop": "40px",
        }
      }
    >
      <button
        className="btn"
        onClick={[Function]}
      >
        Register
      </button>
    </div>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "marginTop": "30px",
      }
    }
  >
    <h6
      className="MuiTypography-root MuiTypography-h6"
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Already have an account?
    </h6>
    <div
      style={
        Object {
          "color": "#355e3b",
        }
      }
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <p
            className="MuiTypography-root MuiTypography-body1"
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Log in
          </p>
        </span>
      </button>
    </div>
  </div>
</div>
`;
